### PR TITLE
Fix #48 by improving handling of Uris with/without "$metadata" suffix.

### LINKED ISSUE #48: Metadata URL with values in Query String
Hi.

The url to get the metadata from my OData API requires setting the Api version in a parameter sent in the url query string.

http://[host]:[port]/odata/$metadata **?api-version=1.0**

In this case, when you try to configure the connected service, an error occurs because the extension tries to concatenate "$metadata" to the end of the URL.

How can I use this feature with the exposed scenario?

Thank you

Victor Perez

image

### PR DESCRIPTION
**Pull request by dr-consit**
Query and fragment parts of the Uri were always clobbered, these are now
preserved, except in the special case where the last segment of the Uri is
"$metadata/".

Please note, that some of the code is perhaps slightly unwieldy, to handle
the "$metadata/" edge-case. This is done to satisfy the specific test-case:
"GetMetadataDocumentUriWithSlashShouldIgnoreQueries".

I am not sure I understand the reasoning behind this test-case, but as
mentioned, the code is made to satisfy this test.

If there is no reason to satisfy this test, that is, if it is OK to always preserve
query parameters, the whole "preserveQueryAndFragment" part of this patch
can be eliminated.

### CODE REVIEW

**Comment by dr-consit** [comment]
This is to fix #48 which is closed but, as far as I can tell, was never actually fixed.

File: src/ODataConnectedService/Common/ExtensionMethods.cs [hunk 1]
```diff
@@ -42,5 +42,40 @@ public static void CopyPropertiesFrom(this object target, object source)
...
+        /// Ensures that the Uri conforms to the expected "$metadata" format
+        /// </summary>
+        /// <param name="uri">Uri to clean</param>
+        /// <returns>Cleaned uri</returns>
+        public static Uri CleanMetadataUri(this Uri uri)
+        {
+            if (uri.Scheme == "http" || uri.Scheme == "https")
+            {
+                UriBuilder uriBuilder;
+                bool preserveQueryAndFragment = true;
```
Comment (KenitoInc) [inline src/ODataConnectedService/Common/ExtensionMethods.cs:57]:
```suggestion
                bool preserveQueryAndFragment = true;
                
```

File: src/ODataConnectedService/Common/ExtensionMethods.cs [hunk 2]
```diff
@@ -42,5 +42,40 @@ public static void CopyPropertiesFrom(this object target, object source)
...
+                if (uri.Segments.Last().StartsWith("$metadata", StringComparison.InvariantCultureIgnoreCase))
+                {
+                    preserveQueryAndFragment = !uri.AbsolutePath.EndsWith("/", StringComparison.Ordinal);
+                    var absolutePathUri = new UriBuilder(uri.Scheme, uri.Host, uri.Port, uri.AbsolutePath.TrimEnd('/')).Uri;
```
Comment (KenitoInc) [inline src/ODataConnectedService/Common/ExtensionMethods.cs:61]:
```suggestion
                    UriBuilder absolutePathUri = new UriBuilder(uri.Scheme, uri.Host, uri.Port, uri.AbsolutePath.TrimEnd('/')).Uri;
```

File: src/ODataConnectedService/Common/ExtensionMethods.cs [hunk 3]
```diff
@@ -42,5 +42,40 @@ public static void CopyPropertiesFrom(this object target, object source)
...
+                    uriBuilder = new UriBuilder(absolutePathUri);
+                }
+                else
+                {
+                    var absolutePathUri = new UriBuilder(uri.Scheme, uri.Host, uri.Port, uri.AbsolutePath.TrimEnd('/') + "/").Uri;
```
Comment (KenitoInc) [inline src/ODataConnectedService/Common/ExtensionMethods.cs:66]:
```suggestion
                    UriBuilder absolutePathUri = new UriBuilder(uri.Scheme, uri.Host, uri.Port, uri.AbsolutePath.TrimEnd('/') + "/").Uri;
```

File: src/ODataConnectedService/Common/ExtensionMethods.cs [hunk 4]
```diff
@@ -42,5 +42,40 @@ public static void CopyPropertiesFrom(this object target, object source)
...
+                    uriBuilder = new UriBuilder(new Uri(absolutePathUri, "$metadata"));
+                }
```
Comment (KenitoInc) [inline src/ODataConnectedService/Common/ExtensionMethods.cs:68]:
```suggestion
                }
                
```

File: src/ODataConnectedService/Common/ExtensionMethods.cs (hunk 1 above)
Comment (KenitoInc) [inline src/ODataConnectedService/Common/ExtensionMethods.cs:57]: Add comments on what this `bool` property means.

File: src/ODataConnectedService/Templates/ODataT4CodeGenerator.ttinclude [hunk 5]
```diff
@@ -159,12 +159,27 @@ public string MetadataDocumentUri
 
         if (uri.Scheme == "http" || uri.Scheme == "https")
         {
-            value = uri.Scheme + "://" + uri.Authority + uri.AbsolutePath;
-            value = value.TrimEnd('/');
-            if (!value.EndsWith("$metadata"))
+            UriBuilder uriBuilder;
+            bool preserveQueryAndFragment = true;
```
Comment (KenitoInc) [inline src/ODataConnectedService/Templates/ODataT4CodeGenerator.ttinclude:163]:
```suggestion
            bool preserveQueryAndFragment = true;
            
```

File: src/ODataConnectedService/Templates/ODataT4CodeGenerator.cs [hunk 6]
```diff
@@ -303,15 +304,9 @@ public string MetadataDocumentUri
...
-        if (uri.Scheme == "http" || uri.Scheme == "https")
-        {
-            value = uri.Scheme + "://" + uri.Authority + uri.AbsolutePath;
-            value = value.TrimEnd('/');
-            if (!value.EndsWith("$metadata"))
-            {
-                value += "/$metadata";
-            }
-        }
+        uri = uri.CleanMetadataUri();
```
Comment (KenitoInc) [inline src/ODataConnectedService/Templates/ODataT4CodeGenerator.cs:307]: It seems you updated this code manually instead of generating it using  `run custom tool` on `ODataT4CodeGenerator.tt` file.

File: src/ODataConnectedService/Templates/ODataT4CodeGenerator.ttinclude [hunk 7]
```diff
@@ -159,12 +159,27 @@ public string MetadataDocumentUri
 
         if (uri.Scheme == "http" || uri.Scheme == "https")
         {
-            value = uri.Scheme + "://" + uri.Authority + uri.AbsolutePath;
-            value = value.TrimEnd('/');
-            if (!value.EndsWith("$metadata"))
+            UriBuilder uriBuilder;
```
Comment (KenitoInc) [inline src/ODataConnectedService/Templates/ODataT4CodeGenerator.ttinclude:162]: This logic is repeated in the `ExtensionMethods.CleanMetadataUri` method.
Kindly refactor by calling` CleanMetadataUri` from here.

**Review by KenitoInc — CHANGES_REQUESTED** [review]

**Comment by KenitoInc** [comment]
@dr-consit  Work on review comments and also rebase the branch to resolve conflicts.

**Comment by KenitoInc** [comment]
> @dr-consit Work on review comments and also rebase the branch to resolve conflicts.

Kind reminder
